datasets/flwr_datasets/partitioner/shard_partitioner.py

```diff
@@ -0,0 +1,322 @@
+# Copyright 2023 Flower Labs GmbH. All Rights Reserved.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+# ==============================================================================
+"""Shard partitioner class."""
+# pylint: disable=R0912
+import math
+from typing import Dict, List, Optional
+
+import numpy as np
+
+import datasets
+from flwr_datasets.partitioner.partitioner import Partitioner
+
+
+class ShardPartitioner(Partitioner):  # pylint: disable=R0902
+    """Partitioner based on shard of (typically) unique classes.
+
+    The algorithm works as follows: the dataset is sorted by label e.g. [samples with
+    label 1, samples with labels 2 ...], then the shards are created, with each
+    shard of size = `shard_size` if provided or automatically calculated:
+    shards_size = len(dataset) / `num_partitions` * `num_shards_per_node`.
+
+    A shard is just a block (part) of a `dataset` that contains `shard_size` consecutive
+    samples. There might be shards that contain samples associated with more than
+    a single unique label. The first case is (remember we have a sorted dataset which
+    is always the prepocessing step) we are at the border between the samples of two
+    classes the shard contains samples of two different classes e.g. the "leftover" of
```

Comment:
```suggestion
    A shard is just a block (chunk) of a `dataset` that contains `shard_size` consecutive
    samples. There might be shards that contain samples associated with more than a
    single unique label. The first case is (remember the preprocessing step sorts the
    dataset by label) when a shard is constructed from samples at the boundaries of the
    sorted dataset and therefore belonging to different classes e.g. the "leftover" of
```

Reply:
I tried rewriting it a bit but i'm not entirely satisfied with it.

Reply:
Sounds good